### PR TITLE
4.0 blog post

### PR DESCRIPTION
**Pull request by Murderlon**

### CODE REVIEW

File: blog/2024-07-3-uppy-4.0.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+slug: 'uppy-4.0'
+published: false
+toc_max_heading_level: 2
+---
+
+Hi ha ho, this is some goofy introduction.
+
+## TypeScript rewrite
+
+In the year 2024 people expect exellent types from their libaries. We used to
```
Comment (lakesare) [inline blog/2024-07-3-uppy-4.0.md:17]: typo `exellent => excellent`

File: blog/2024-07-3-uppy-4.0.md [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+	meta: { myCustomMetadata: 'foo' },
+});
+
+// This is now typed
+const { myCustomMetadata } = uppy.getFile(id).meta;
+
+await uppy.upload();
+
+// This is strictly typed too
+const { someThingMyBackendReturns } = uppy.getFile(id).response;
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:63]: Getting the following TS errors:

```
index.ts:3:23 - error TS2307: Cannot find module '@uppy/xhr-upload' or its corresponding type declarations.

3 import XHRUpload from '@uppy/xhr-upload';
                        ~~~~~~~~~~~~~~~~~~

index.ts:26:9 - error TS2339: Property 'someThingMyBackendReturns' does not exist on type '{ body?: Body; status: number; bytesUploaded?: number; uploadURL?: string; }'.

26 const { someThingMyBackendReturns } = uppy.getFile(id).response;
           ~~~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors in the same file, starting at: index.ts:3
```

Comment (Murderlon) [reply blog/2024-07-3-uppy-4.0.md:63]: Where? CI is green.

Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:63]:
```suggestion
import Uppy from '@uppy/core';
// xhr-upload is for uploading to your own backend.
import XHRUpload from '@uppy/xhr-upload';

// Your own metadata on files
type Meta = { myCustomMetadata: string };
// The response from your server
type Body = { someThingMyBackendReturns: string };

const uppy = new Uppy<Meta, Body>().use(XHRUpload, {
	endpoint: '/upload',
});

const id = uppy.addFile({
	name: 'example.jpg',
	data: new Blob(),
	meta: { myCustomMetadata: 'foo' },
});

// This is now typed
const { myCustomMetadata } = uppy.getFile(id).meta;

await uppy.upload();

// This is strictly typed too
const { someThingMyBackendReturns } = uppy.getFile(id).response.body!;
```

Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:63]:
```suggestion
const { someThingMyBackendReturns } = uppy.getFile(id).response.body!;
```

Comment (aduh95) [reply blog/2024-07-3-uppy-4.0.md:63]: Fix is https://github.com/transloadit/uppy.io/pull/258/files#r1670167026

File: blog/2024-07-3-uppy-4.0.md [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+const [uppy] = useState(() => new Uppy());
+
+const files = useUppyState(uppy, (state) => state.files);
+const totalProgress = useUppyState(uppy, (state) => state.totalProgress);
+// We can also get specific plugin state.
+// Note that the value on `plugins` depends on the `id` of the plugin.
+const metaFields = useUppyState(
+	uppy,
+	(state) => state.plugins?.Dashboard?.metaFields,
+);
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:133]: Getting the following TS errors:

```
$ npx tsc --target esnext --module nodenext --noEmit index.ts                                  
node_modules/@uppy/dashboard/lib/Dashboard.d.ts:3:39 - error TS2307: Cannot find module '@uppy/core/lib/BasePlugin' or its corresponding type declarations.

3 import type { DefinePluginOpts } from '@uppy/core/lib/BasePlugin';
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/drag-drop/lib/DragDrop.d.ts:2:39 - error TS2307: Cannot find module '@uppy/core/lib/BasePlugin' or its corresponding type declarations.

2 import type { DefinePluginOpts } from '@uppy/core/lib/BasePlugin';
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/drag-drop/lib/DragDrop.d.ts:3:38 - error TS2307: Cannot find module '@uppy/core/lib/UIPlugin' or its corresponding type declarations.

3 import type { UIPluginOptions } from '@uppy/core/lib/UIPlugin';
                                       ~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/progress-bar/lib/ProgressBar.d.ts:3:34 - error TS2307: Cannot find module '@uppy/core/lib/Uppy' or its corresponding type declarations.

3 import type { Uppy, State } from '@uppy/core/lib/Uppy';
                                   ~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/provider-views/lib/ProviderView/ProviderView.d.ts:4:36 - error TS2307: Cannot find module '@uppy/utils/lib/CompanionFile.ts' or its corresponding type declarations.

4 import type { CompanionFile } from '@uppy/utils/lib/CompanionFile.ts';
                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/react/lib/Dashboard.d.ts:7:16 - error TS2833: Cannot find namespace 'React'. Did you mean 'preact'?

7 }, 'inline'> & React.BaseHTMLAttributes<HTMLDivElement>;
                 ~~~~~

node_modules/@uppy/react/lib/Dashboard.d.ts:19:51 - error TS2339: Property 'props' does not exist on type 'Dashboard<M, B>'.

19     componentDidUpdate(prevProps: Dashboard<M, B>['props']): void;
                                                     ~~~~~~~

node_modules/@uppy/react/lib/DashboardModal.d.ts:7:40 - error TS2833: Cannot find namespace 'React'. Did you mean 'preact'?

7 }, 'inline' | 'onRequestCloseModal'> & React.BaseHTMLAttributes<HTMLDivElement>;
                                         ~~~~~

node_modules/@uppy/react/lib/DashboardModal.d.ts:25:56 - error TS2339: Property 'props' does not exist on type 'DashboardModal<M, B>'.

25     componentDidUpdate(prevProps: DashboardModal<M, B>['props']): void;
                                                          ~~~~~~~

node_modules/@uppy/react/lib/DragDrop.d.ts:16:50 - error TS2339: Property 'props' does not exist on type 'DragDrop<M, B>'.

16     componentDidUpdate(prevProps: DragDrop<M, B>['props']): void;
                                                    ~~~~~~~

node_modules/@uppy/react/lib/ProgressBar.d.ts:15:53 - error TS2339: Property 'props' does not exist on type 'ProgressBar<M, B>'.

15     componentDidUpdate(prevProps: ProgressBar<M, B>['props']): void;
                                                       ~~~~~~~

node_modules/@uppy/react/lib/StatusBar.d.ts:16:51 - error TS2339: Property 'props' does not exist on type 'StatusBar<M, B>'.

16     componentDidUpdate(prevProps: StatusBar<M, B>['props']): void;
                                                     ~~~~~~~

node_modules/@uppy/status-bar/lib/StatusBar.d.ts:3:34 - error TS2307: Cannot find module '@uppy/core/lib/Uppy' or its corresponding type declarations.

3 import type { Uppy, State } from '@uppy/core/lib/Uppy';
                                   ~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/status-bar/lib/StatusBar.d.ts:4:39 - error TS2307: Cannot find module '@uppy/core/lib/BasePlugin' or its corresponding type declarations.

4 import type { DefinePluginOpts } from '@uppy/core/lib/BasePlugin';
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@uppy/status-bar/lib/StatusBarOptions.d.ts:1:38 - error TS2307: Cannot find module '@uppy/core/lib/UIPlugin' or its corresponding type declarations.

1 import type { UIPluginOptions } from '@uppy/core/lib/UIPlugin';
                                       ~~~~~~~~~~~~~~~~~~~~~~~~~


Found 15 errors in 11 files.

Errors  Files
     1  node_modules/@uppy/dashboard/lib/Dashboard.d.ts:3
     2  node_modules/@uppy/drag-drop/lib/DragDrop.d.ts:2
     1  node_modules/@uppy/progress-bar/lib/ProgressBar.d.ts:3
     1  node_modules/@uppy/provider-views/lib/ProviderView/ProviderView.d.ts:4
     2  node_modules/@uppy/react/lib/Dashboard.d.ts:7
     2  node_modules/@uppy/react/lib/DashboardModal.d.ts:7
     1  node_modules/@uppy/react/lib/DragDrop.d.ts:16
     1  node_modules/@uppy/react/lib/ProgressBar.d.ts:15
     1  node_modules/@uppy/react/lib/StatusBar.d.ts:16
     2  node_modules/@uppy/status-bar/lib/StatusBar.d.ts:3
     1  node_modules/@uppy/status-bar/lib/StatusBarOptions.d.ts:1
```

File: blog/2024-07-3-uppy-4.0.md [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+### `useUppyState(uppy, selector)`
+
+Use this hook when you need to access Uppy’s state reactively.
+
+```js
+import Uppy from '@uppy/core';
+import { useUppyState } from '@uppy/react';
+
+// IMPORTANT: passing an initializer function to prevent Uppy from being recreated on every render.
+const [uppy] = useState(() => new Uppy());
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:124]:
```suggestion
import {useState} from 'react'
const [uppy] = useState(() => new Uppy());
```

File: blog/2024-07-3-uppy-4.0.md [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+your use case, you may want to keep the values in state or clear the state after
+something else happened.
+
+```ts
+import Uppy from '@uppy/core';
+import { useUppyEvent } from '@uppy/react';
+
+// IMPORTANT: passing an initializer function
+// to prevent Uppy from being recreated on every render.
+const [uppy] = useState(() => new Uppy());
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:161]:
```suggestion
import { useState } from 'react';
const [uppy] = useState(() => new Uppy());
```

File: blog/2024-07-3-uppy-4.0.md (hunk 3 above)
Comment (Murderlon) [reply blog/2024-07-3-uppy-4.0.md:133]: Probably better to test how this is actually used, so inside a component and with the right tsconfig.

https://stackblitz.com/edit/vitejs-vite-vehvbq?file=src%2FApp.tsx

File: blog/2024-07-3-uppy-4.0.md [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+**Not using
+[Signature Authentication](https://transloadit.com/docs/topics/signature-authentication/)
+can be a security risk**. Signature Authentication is a security measure that
+can prevent outsiders from tampering with your Assembly Instructions.
+
+This means the majority of implementers will write something like this, as
+recommended:
+
+```ts
+new Uppy().use(Transloadit, {
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:302]:
```suggestion
import { Uppy } from '@uppy/core'
import Transloadit from '@uppy/transloadit'

new Uppy().use(Transloadit, {
```

File: blog/2024-07-3-uppy-4.0.md [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+	async assemblyOptions(file) {
+		const res = await fetch('/transloadit-params');
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:304]:
```suggestion
		const response = await fetch('/transloadit-params');
```

File: blog/2024-07-3-uppy-4.0.md [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+		return response.json();
+	},
+});
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:307]: Getting the following TS error:

```
npx tsc --target esnext --module nodenext --noEmit index.ts
index.ts:5:8 - error TS2322: Type '(file: any) => Promise<any>' is not assignable to type 'AssemblyOptions | (() => AssemblyOptions | Promise<AssemblyOptions>)'.
  Type '(file: any) => Promise<any>' is not assignable to type '() => AssemblyOptions | Promise<AssemblyOptions>'.
    Target signature provides too few arguments. Expected 1 or more, but got 0.

5  async assemblyOptions(file) {
         ~~~~~~~~~~~~~~~

  node_modules/@uppy/transloadit/lib/index.d.ts:120:5
    120     assemblyOptions?: AssemblyOptions | (() => Promise<AssemblyOptions> | AssemblyOptions);
            ~~~~~~~~~~~~~~~
    The expected type comes from property 'assemblyOptions' which is declared here on type 'TransloaditOptions<Meta, Body>'


Found 1 error in index.ts:5
```

File: blog/2024-07-3-uppy-4.0.md [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+
+```ts
+import Uppy from '@uppy/core';
+import { useUppyEvent } from '@uppy/react';
+
+// IMPORTANT: passing an initializer function
+// to prevent Uppy from being recreated on every render.
+const [uppy] = useState(() => new Uppy());
+
+const [results, clearResults] = useUppyEvent(uppy, 'transloadit:result');
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:163]: Getting the following TS error:

```
index.ts:8:52 - error TS2345: Argument of type '"transloadit:result"' is not assignable to parameter of type 'keyof UppyEventMap<M, B>'.

8 const [results, clearResults] = useUppyEvent(uppy, 'transloadit:result');
                                                     ~~~~~~~~~~~~~~~~~~~~
```

File: blog/2024-07-3-uppy-4.0.md [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+const [stepName, result, assembly] = results; // strongly typed
+
+useUppyEvent(uppy, 'cancel-all', doSomethingElse);
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:166]:
```suggestion
useUppyEvent(uppy, 'cancel-all', () => {
  // Handle event here.
});
```

**Comment by lakesare** [comment]
@Murderlon, uploaded the videos & reworked the provider-views description a bit.
I removed the updates that are more bug fixes than new functionality (e.g. "made Unsplash display more than one page", "fixed shift-clicking troubles", "fixed the way nOfSelectedFiles works"). Tell me if you think we should readd them to the post (I'll do it).

I'll upload slightly adjusted videos shortly, will cut a few seconds off the end of each video (it shows the border).

**Comment by lakesare** [comment]
Done.
Please edit my text as you see fit.

**Comment by Murderlon** [comment]
Looks great, thanks!

File: blog/2024-07-3-uppy-4.0.md
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:1]: The date should be 2024-07-10, when the release was pushed to npm

File: blog/2024-07-3-uppy-4.0.md [hunk 11]
```diff
@@ -0,0 +1,408 @@
+---
+title:
+  'New Uppy 4.0 major: TypeScript rewrite, Google Photos, React hooks, and much
+  more.'
+date: 2024-07-03
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:5]: The date should be 2024-07-10, when the release was pushed to npm
```suggestion
date: 2024-07-10
```
Also the name of the file needs to be amended

File: blog/2024-07-3-uppy-4.0.md [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+- Begin an upload before you know the final object size – You can upload an
+  object as you are creating it.
+
+However, the downside is request overhead, as it needs to do creation, signing,
+and completion requests besides the upload requests. For example, if you are
+uploading files that are only a couple kilobytes with a 100ms roundtrip latency,
+you are spending 400ms on overhead and only a few milliseconds on uploading.
+This really adds up if you upload a lot of small files.
+
+AWS, and generally the internet from what we found, tend to agree that **you
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md:98]:
```suggestion
AWS docs, and generally the internet from what we found, tend to agree that **you
```

Comment (Murderlon) [reply blog/2024-07-3-uppy-4.0.md]: not an improvement I'd say

File: blog/2024-07-3-uppy-4.0.md [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+
+```ts
+import type { State } from '@uppy/core';
+```
+
+### `useUppyEvent(uppy, event, callback)`
+
+Listen to Uppy [events](/docs/uppy/#events) in a React component.
+
+Returns an array of which the first item is an array of results from the event.
```
Comment (aduh95) [inline blog/2024-07-3-uppy-4.0.md]: Not sure how to re-phrase that, but I find the current wording hard to parse.

- maybe instead of array, we could say pair?
- "the first item is an array of results from the event" – but only if an event has been emitted

File: blog/2024-07-10-uppy-4.0.md [hunk 14]
```diff
@@ -0,0 +1,421 @@
...
+
+// This is strictly typed too
+const { someThingMyBackendReturns } = uppy.getFile(id).response.body!;
+```
+
+Happy inferring!
+
+## Merging the two AWS S3 plugins
+
+We used to two separate plugins for uploading to S3 (and S3-compatible
```
Comment (aduh95) [inline blog/2024-07-10-uppy-4.0.md:89]:
```suggestion
We used to have two separate plugins for uploading to S3 (and S3-compatible
```

File: blog/2024-07-10-uppy-4.0.md [hunk 15]
```diff
@@ -0,0 +1,421 @@
...
+  you initiate a multipart upload, there is no expiry; you must explicitly
+  complete or stop the multipart upload.
+- Begin an upload before you know the final object size – You can upload an
+  object as you are creating it.
+
+However, the downside is request overhead, as it needs to do creation, signing,
+and completion requests besides the upload requests. For example, if you are
+uploading files that are only a couple kilobytes with a 100ms roundtrip latency,
+you are spending 400ms on overhead and only a few milliseconds on uploading.
+This really adds up if you upload a lot of small files.
```
Comment (aduh95) [inline blog/2024-07-10-uppy-4.0.md:106]: nit: I wonder if we'd want to present the info in a table that shows (one column for advantages, one column for disadvantages)

File: blog/2024-07-10-uppy-4.0.md [hunk 16]
```diff
@@ -0,0 +1,421 @@
...
+import type { State } from '@uppy/core';
+```
+
+### `useUppyEvent(uppy, event, callback)`
+
+Listen to Uppy [events](/docs/uppy/#events) in a React component.
+
+The hook returns `[results, clear]`. `results` is an array of values from the
+event. Depending on the event, that can be empty or have up to three values.
+`clear` is a function to clear the state.
```
Comment (aduh95) [inline blog/2024-07-10-uppy-4.0.md:170]: nit: I don't find the phrasing "clear the state" very clear
```suggestion
`clear` is a function that, when called, clears the `results` array.
```

File: blog/2024-07-10-uppy-4.0.md [hunk 17]
```diff
@@ -0,0 +1,421 @@
...
+
+Values remain in state until the next event (if that ever comes). Depending on
```
Comment (aduh95) [inline blog/2024-07-10-uppy-4.0.md:172]:
```suggestion
Values remain in state until the next event (if that ever comes) – or if the state is manually cleared. Depending on
```

File: blog/2024-07-10-uppy-4.0.md [hunk 18]
```diff
@@ -0,0 +1,421 @@
...
+:::
+
+<video src="/img/blog/4.0/google-photos.mp4" controls></video>
+
+[`@uppy/google-photos`](/docs/google-photos/) is a new plugin so you can use it
+next to your existing [`@uppy/google-drive`](/docs/google-drive/) plugin.
+
+## UX improvements for viewing remote files
+
+When using [Dashboard](/docs/dashboard) with any of our remote sources ( Google
```
Comment (aduh95) [inline blog/2024-07-10-uppy-4.0.md:218]:
```suggestion
When using [Dashboard](/docs/dashboard) with any of our remote sources (Google
```

File: blog/2024-07-10-uppy-4.0.md [hunk 19]
```diff
@@ -0,0 +1,421 @@
...
+Drive, Facebook, etc.) you use our internal `@uppy/provider-views` plugin to
```
Comment (aduh95) [inline blog/2024-07-10-uppy-4.0.md:219]:
```suggestion
Drive, Facebook, etc.), you use our internal `@uppy/provider-views` plugin to
```

File: blog/2024-07-10-uppy-4.0.md [hunk 20]
```diff
@@ -0,0 +1,421 @@
...
+With streaming upload disabled, the whole file will be downloaded first. The
+upload will then start when the download has completely finished.
+
+When streaming upload is enabled, Companion will start downloading the file from
+the provider (such as Google Drive), while at the same time starting the upload
+to the destination (such as Tus), and sending every chunk of data consecutively.
+
+For more information, see the [Companion docs](/docs/companion/).
+
+### New required option `corsOrigins`
```
Comment (aduh95) [inline blog/2024-07-10-uppy-4.0.md:405]:
```suggestion
### `corsOrigins` option is now required
```

File: blog/2024-07-10-uppy-4.0.md [hunk 21]
```diff
@@ -0,0 +1,421 @@
...
+
+As a security measure, we now require the
+[`corsOrigins`](/docs/companion/#corsorigins) option to be set.
+
+It serves two purposes, it sets the `Access-Control-Allow-Origin` header and it
+sets the origin for `window.postMessage()`, which is needed to communicate the
```
Comment (aduh95) [inline blog/2024-07-10-uppy-4.0.md:411]:
```suggestion
sets the target origin for `window.postMessage()`, which is needed to communicate the
```

File: blog/2024-07-10-uppy-4.0.md (hunk 15 above)
Comment (Murderlon) [reply blog/2024-07-10-uppy-4.0.md:106]: I think it's fine like this. We're also only mentioning one downside so it would be a very uneven table

File: blog/2024-07-10-uppy-4.0.md [hunk 22]
```diff
@@ -0,0 +1,422 @@
...
+
+<img src="/img/blog/4.0/preview.jpg" alt="Screenshot of Uppy Dashboad with text outlining the major new features in 4.0" />
+
+Hold on to your leashes, folks! Uppy 4.0 is here, and it’s more exciting than a
+tennis ball at the dog park! Our beloved Uppy mascot, the adorable coding
+canine, has been hard at work fetching all the latest updates for you.
+
+From a full TypeScript makeover to shiny new React hooks, and even Google Photos
+integration, this release is packed with treats that will make developers wag
+their tails in delight. So, sit, stay, and let’s dig into what makes Uppy 4.0
```
Comment (lakesare) [inline blog/2024-07-10-uppy-4.0.md:21]: Lmao this is all so confusing, who's a dog!
First we're saying that Uppy is a tennis ball at the dog park.
Then we're saying that Uppy is a dog.
Then we're saying that developers are dogs who wag their tails.

I'm pro this vibe, but I think we could be more focused on who is a dog in this scenario.

Comment (lakesare) [reply blog/2024-07-10-uppy-4.0.md:21]: I think the idea was that Uppy is a dog whose name is Uppy

File: blog/2024-07-10-uppy-4.0.md [hunk 23]
```diff
@@ -0,0 +1,422 @@
...
+a `boolean` or a function to determine it per file.
+
+## React hooks
+
+People working with React are more likely to create their own user interface on
+top of Uppy than those working with “vanilla” setups. Working with our pre-build
+UI components is a plug-and-play experience, but building on top of Uppy’s state
+with React primitives has been tedious.
+
+To address this we’re introducing to new hooks: `useUppyState` and
```
Comment (lakesare) [inline blog/2024-07-10-uppy-4.0.md:128]: Typo: `to` => `two`

File: blog/2024-07-10-uppy-4.0.md [hunk 24]
```diff
@@ -0,0 +1,422 @@
...
+<video src="/img/blog/4.0/google-photos.mp4" controls></video>
+
+[`@uppy/google-photos`](/docs/google-photos/) is a new plugin so you can use it
+next to your existing [`@uppy/google-drive`](/docs/google-drive/) plugin.
+
+## UX improvements for viewing remote files
+
+When using [Dashboard](/docs/dashboard) with any of our remote sources (Google
+Drive, Facebook, etc.) you use our internal `@uppy/provider-views` plugin to
+navigate and select files. In Uppy 4.0, we are making a handful of quality of
```
Comment (lakesare) [inline blog/2024-07-10-uppy-4.0.md:220]:
```suggestion
navigate the folders and select files. In Uppy 4.0, we are making a handful of quality of
```

File: blog/2024-07-10-uppy-4.0.md [hunk 25]
```diff
@@ -0,0 +1,422 @@
...
+    <tr>
+      <td style="width: 50%;"><video src="/img/blog/4.0/restrictions-old.mp4" controls></video></td>
+      <td style="width: 50%;"><video src="/img/blog/4.0/restrictions-new.mp4" controls></video></td>
+    </tr>
+  </tbody>
+</table>
+
+We’re confident this turns our interface for remote sources into the most
+advanced one out there. We’ve seen some competing libraries not even aggregating
+results beyond the first page API limit of providers.
```
Comment (lakesare) [inline blog/2024-07-10-uppy-4.0.md:275]:
```suggestion
results beyond the first page returned by the provider API.
```

File: blog/2024-07-10-uppy-4.0.md [hunk 26]
```diff
@@ -0,0 +1,422 @@
...
+
+## Revamped options for `@uppy/xhr-upload`
+
+Before the plugin had the options `getResponseData`, `getResponseError`,
```
Comment (lakesare) [inline blog/2024-07-10-uppy-4.0.md:279]:
```suggestion
In previous versions, the `@uppy/xhr-upload` plugin had the options `getResponseData`, `getResponseError`,
```

File: blog/2024-07-10-uppy-4.0.md [hunk 27]
```diff
@@ -0,0 +1,422 @@
...
+`validateStatus` and `responseUrlFieldName`. These were inflexible and too
+specific. Now we have hooks similar to `@uppy/tus`:
+
+- `onBeforeRequest` to manipulate the request before it is sent.
+- `shouldRetry` to determine if a request should be retried. By default 3
+  retries with exponential backoff. After three attempts it will throw an error,
+  regardless of whether you returned `true`.
+- `onAfterResponse` called after a successful response but before Uppy resolves
```
Comment (lakesare) [inline blog/2024-07-10-uppy-4.0.md:287]:
```suggestion
- `onAfterResponse` called after a successful response, but before Uppy resolves
```

File: blog/2024-07-10-uppy-4.0.md (hunk 22 above)
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:21]: Agreed. The way I pretty much saw it when I helped conceive this idea is:

- Uppy, and to a certain extent the developers / blog writers behind it, can be attributed doglike qualities
- users of Uppy / readers of the blogs can be depicted as dog owners

I slightly changed up the intro in my PR. Couldn't help myself from adding a dumb pun. (In my mind, 'without further a-dog' = no more dog jokes, for now).

File: blog/2024-07-10-uppy-4.0.md [hunk 28]
```diff
@@ -0,0 +1,422 @@
...
+UI components is a plug-and-play experience, but building on top of Uppy’s state
+with React primitives has been tedious.
+
+To address this we’re introducing to new hooks: `useUppyState` and
+`useUppyEvent`. Thanks to the TypeScript rewrite, we can now do powerful
+inference in hooks as well.
+
+### `useUppyState(uppy, selector)`
+
+Use this hook when you need to access Uppy’s state reactively.
```
[the remark this answers is not in the document]
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:134]: Changed this in my PR

File: blog/2024-07-10-uppy-4.0.md [hunk 29]
```diff
@@ -0,0 +1,422 @@
...
+// Note that the value on `plugins` depends on the `id` of the plugin.
+const metaFields = useUppyState(
+	uppy,
+	(state) => state.plugins?.Dashboard?.metaFields,
+);
+```
+
+You can see all the values you can access on the
+[`State`](https://github.com/transloadit/uppy/blob/dab8082a4e67c3e7f109eacfbd6c3185f117dc60/packages/%40uppy/core/src/Uppy.ts#L156)
+type. If you are accessing plugin state, you would have to look at the types of
```
[the remark this answers is not in the document]
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:157]: Changed the text to your suggestion in my PR. 

I did not remove any code, since I feel it is better that I never do that.

File: blog/2024-07-10-uppy-4.0.md [hunk 30]
```diff
@@ -0,0 +1,422 @@
...
+
+This means a 5GB video isn’t eating into your users’ data plans and you don’t
+have to worry about OAuth.
+
+:::
+
+<video src="/img/blog/4.0/google-photos.mp4" controls></video>
+
+[`@uppy/google-photos`](/docs/google-photos/) is a new plugin so you can use it
+next to your existing [`@uppy/google-drive`](/docs/google-drive/) plugin.
```
[the remark this answers is not in the document]
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:214]: Made this change in my PR, with a slight alteration.

File: blog/2024-07-10-uppy-4.0.md [hunk 31]
```diff
@@ -0,0 +1,422 @@
...
+	},
+	async onAfterResponse(xhr) {
+		if (xhr.status === 401) {
+			token = await getAuthToken();
+		}
+	},
+});
+```
+
+Checkout the `@uppy/xhr-upload` [docs](/docs/xhr-upload/) for more info.
```
[the remark this answers is not in the document]
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:321]: Fixed in my PR

File: blog/2024-07-10-uppy-4.0.md [hunk 32]
```diff
@@ -0,0 +1,422 @@
...
+  _per file_.
+
+You can read more about Assembly Variables in the
+[docs](https://transloadit.com/docs/topics/assembly-variables/).
+
+## Companion
+
+### Streaming uploads by default
+
+Streaming uploads are now the default in in Companion. This comes with greatly
```
[the remark this answers is not in the document]
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:388]: Fixed in my PR

File: blog/2024-07-10-uppy-4.0.md [hunk 33]
```diff
@@ -0,0 +1,422 @@
...
+### `corsOrigins` option is now required
+
+As a security measure, we now require the
+[`corsOrigins`](/docs/companion/#corsorigins) option to be set.
+
+It serves two purposes, it sets the `Access-Control-Allow-Origin` header and it
+sets the target origin for `window.postMessage()`, which is needed to
+communicate the OAuth token from the new tab you used to log-in to a provider
+back to Companion.
+
```
[the remark this answers is not in the document]
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:414]: Just noticed that a few of your comments were initially hidden, so I missed them during my iteration.

I made a sort-of similar change to this sentence in my PR, only getting rid of the 2x 'it sets'.

I am not clear on the final part of the sentence either and, to be honest, your version doesn't completely make sense to me either. I left a comment in my PR, so I'll leave this for someone else to clear up.

File: blog/2024-07-10-uppy-4.0.md [hunk 34]
```diff
@@ -0,0 +1,422 @@
...
+## And more
+
+The 4.0 release contained over 170 contributions, many too small to mention, but
+together resulting in Uppy continuing to grow and improve. We closely listen to
+the community and are always looking for ways to improve the experience.
```
[the remark this answers is not in the document]
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:419]: Made this change in my PR

File: blog/2024-07-10-uppy-4.0.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:128]: Fixed in my PR

File: blog/2024-07-10-uppy-4.0.md [hunk 35]
```diff
@@ -0,0 +1,422 @@
...
+
+## Google Photos
+
+A long requested feature is finally here: Google Photos support!
+
+:::info
+
+Uppy can bring in files from the cloud with [Companion](/docs/companion/).
+
+Companion is a hosted, standalone, or middleware server to take away the
```
[the remark this answers is not in the document]
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:202]: Fixed in my PR

File: blog/2024-07-10-uppy-4.0.md (hunk 24 above)
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:220]: Made this change in my PR

File: blog/2024-07-10-uppy-4.0.md (hunk 25 above)
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:275]: Nice! This part of the sentence was not making sense to me, but I scratched it up to me simply being ignorant about terminology.

Made this change in my PR

File: blog/2024-07-10-uppy-4.0.md (hunk 26 above)
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:279]: Made this change in my PR

File: blog/2024-07-10-uppy-4.0.md (hunk 27 above)
Comment (AJvanLoon) [reply blog/2024-07-10-uppy-4.0.md:287]: Fixed in my PR
